README.md
# SilverTech Inc. HTML5 Boilerplate
The SilverTech Inc. Boilerplate was created using the HTML5 Boilerplate.  It is not intended for distribution outside SilverTech.  This boilerplate serves as a template to speed up the development process.  The folder structure loosely matches the structure of SilverTech's CMS install to allow for easy transfer to .ASP upon completion of the cut up process. 

:warning: We have set up a fairly specific branching model for this repo, please review [here](http://nvie.com/posts/a-successful-git-branching-model/).

:warning:  When comitting changes to the git repo, please remember to update the changelog.md with the date, your name, and changes made. Newest entries to the changelog at the top. 

:warning: Requires node.js.  Download and install node [here](http://nodejs.org/)

## Structure and Features

### Root and General
* The package.json and Gruntfile.js are included and preloaded with the current packages the front end team has identified should be used with every project.  
* server.js is included so that you may (optionally) run a local apache server without the need for WAMP or equivilent software.  Simply  navigate to the project folder in the command prompt and run `node server.js`  you can then navigate to your project at localhost:8000
* the .gitignore is already ignoring compiled sass and javascript.  Grunt will create the compiled files for you.
*

### Javascript 
* respond.js and selectivizr.js are included in the project folder, they are not linked in the markup.  You should know before starting a project whether or not you'll need these polyfills. 

### Sass and CSS
* grunt-criticalcss will automatically inline your above the fold styles.  

:warning: You must run server BEFORE running criticalcss.  It's crawling over the live page as our partials make it impossible to accurately crawl any single html file

### Images

changelog.md
# Changelog

## September 6, 2014 
* Sean Dempsey - Replaced webfont task in Gruntfile.js (went missing somehow).  Added grunt-criticicalcss which currently outputs to partials/criticalcss.html  This is then included in base.html in the '<head></head>' #perf

## August 15, 2014 
* Sean Dempsey - Updated Gruntfile.js css tasks to separate 'development' and 'production' tasks.
  - The task now creates a .min.css file for production and leaves the uncleaned, unminified source for mapping.
* Sean Dempsey - Updated .gitignore to cover all map and .min.css files.

## August 5, 2014
* Sean Dempsey - Updated Gruntfile.js 'imagemin' path from images to 'img'.
* Sean Dempsey - Updated Changelog to reflect changes made.  

:warning: Please don't forget!  You could just copy your intended commit message!  

## July 30, 2014 
* Sean Dempsey - Added grunt-express-server, Separated out mixins, added timing functions (decidedly not a mixin - we can move it later)

##July 27, 2014
* Sean Dempsey - Updated _base.scss with triangle mixin.

## July 24, 2014
* Matthew Chase - Got rid of the HTML5 bootstrap index.html

## July 23, 2014
* Matthew Chase - Updates Grunt to include grunt-githooks and grunt-npm-install, which (if this works) will invoke npm install and grunt after you update the repo.

* Matthew Chase - Removed link tag trying to load normalize.css

## July 18, 2014
* Matthew Chase - Modified swig templates so have a base template and page layouts that extend it. Now inner has two blocks; one for content and one for sidebar, instead of just one block that contains ARTICLE and ASIDE tags
* Matthew Chase - Added Swig templating and a bare-bones icon font

##July 3, 2014
* Sean Dempsey - Created!



index.js
var express = require('express'),
    bodyParser = require('body-parser'),
    path = require('path'),
    fs = require('fs'),
    request = require('request'),
    favicon = require('serve-favicon');
    logger = require('morgan');
    app = express();

var port = process.env.PORT || 1337;
var host = '0.0.0.0';

var faviconPath = '/cutup/public/library/img/common/favicon.ico';

// app.use(logger());

var ztatic = process.argv[2] === 'static';
if (ztatic) {
    // Serve up the static version of the site instead of the express version.
    if (fs.existsSync(__dirname + faviconPath)) {
        app.use(favicon(__dirname + faviconPath));
    }
    app.use('/', express.static(__dirname + '/public'));
    app.listen(port, host, function () {
        console.log("Static site listening on host: " + host + " and port: " + port);
    });
}
else {
    // parse application/x-www-form-urlencoded
    app.use(bodyParser.urlencoded({ extended: false }))
    // parse application/json
    app.use(bodyParser.json())
    
    app.set('view cache', false);

    if (fs.existsSync(__dirname + faviconPath)) {
        app.use(favicon(__dirname + faviconPath));
    }
    require('./cutup/routes')(app);
    app.use('/', express.static(__dirname + '/cutup/public'));

    var server = app.listen(port, host, function () {
        console.log("Listening on host: " + host + " and port: " + port);

        var crawl = process.argv[2] === 'crawl';
        var outputFolder = path.join(__dirname, "/cutup/public");
        fs.mkdir(outputFolder, function (err) { });
        if (crawl) { 
            crawlApp('localhost', port, server, app, outputFolder);
        }
    });
}

//
// Crawl the app and save the static files.
//
function crawlApp(host, port, server, app, outputFolder) {
    var stack = app._router.stack;
    for (var i in stack) {
        var rt = stack[i];

        if (rt.route) {
            var pth = rt.route.path;
            if (typeof(pth) !== 'string') {
                continue;
            }
            // Hit this path and get the HTML response.
            var url = "http://" + host + ":" + port;
            var name = rt.route.path.replace(/\//, '');
            if (!name) { 
                name = "index"; 
            }
            name += ".html";

            var fileToWrite = path.join(outputFolder, name);
            var fullPath = url + rt.route.path;
            
            // Wrap this in a closure so that the value of the path and filename to write don't change with the next iteration of the loop by the time the HTTP request comes back.
            (function (route, fullPath, fileToWrite) {
                // Save the response to disk.
                console.log("Getting " + fullPath);
                console.log("Output file: " + fileToWrite);

                request.get(fullPath, function (err, res, body) {
                    body = body.replace(/href=['"](\/[a-z\-]+)['"]/gi, 'href=\'$1.html\'');
                    body = body.replace(/href=['"]\/['"]/gi, 'href="/index.html"');

                    fs.writeFile(fileToWrite, body, function (err) {
                        console.log("Wrote ", fileToWrite);
                    });
                });

            })(rt.route, fullPath, fileToWrite);
        }
    }
}


crawler.js
var express = require('express'),
    bodyParser = require('body-parser'),
    path = require('path'),
    fs = require('fs'),
    request = require('request'),
    favicon = require('serve-favicon'),
    logger = require('morgan'),
    FeedParser = require('feedparser'),
    ogs = require('open-graph-scraper'),
    app = express();
    require("string_score");


var port = process.env.PORT || 8080;
var host = '0.0.0.0';
var ogdata = [];
var router = express.Router();              // get an instance of the express Router

function mergeFeed(feed){
	ogdata.push(feed)
}

function fetch(feed) {
  // Define our streams
  var req = request(feed, {timeout: 10000, pool: false});
  req.setMaxListeners(50);
  // Some feeds do not respond without user-agent and accept headers.
  req.setHeader('user-agent', 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_8_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/31.0.1650.63 Safari/537.36')
     .setHeader('accept', 'text/html,application/xhtml+xml');

  var feedparser = new FeedParser();

  // Define our handlers
  req.on('error', done);
  req.on('response', function(res) {
    if (res.statusCode != 200) return this.emit('error', new Error('Bad status code'));
    var charset = getParams(res.headers['content-type'] || '').charset;
    res = maybeTranslate(res, charset);
    // And boom goes the dynamite
    res.pipe(feedparser);
  });

  feedparser.on('error', done);
  feedparser.on('end', done);
  feedparser.on('readable', function() {
    var post;
    while (post = this.read()) {
		var options = {'url':post.origlink,'timeout':2000};
		ogs(options, function(err, results) {
			if (err === null){
			    console.log("err:",err);
			    console.log("results:",results.data);
			    if (ogdata.length > 0){
			    	mergeFeed(results.data)
			    }else{
			    	ogdata.push(results.data);
				}
			}
		});
      	//console.log(post.origlink);
    }
  });
  return ogdata;
}

function maybeTranslate (res, charset) {
  var iconv;
  // Use iconv if its not utf8 already.
  if (!iconv && charset && !/utf-*8/i.test(charset)) {
    try {
      iconv = new Iconv(charset, 'utf-8');
      console.log('Converting from charset %s to utf-8', charset);
      iconv.on('error', done);
      // If we're using iconv, stream will be the output of iconv
      // otherwise it will remain the output of request
      res = res.pipe(iconv);
    } catch(err) {
      res.emit('error', err);
    }
  }
  return res;
}

function getParams(str) {
  var params = str.split(';').reduce(function (params, param) {
    var parts = param.split('=').map(function (part) { return part.trim(); });
    if (parts.length === 2) {
      params[parts[0]] = parts[1];
    }
    return params;
  }, {});
  return params;
}

function done(err) {
  if (err) {
    console.log(err, err.stack);
    return process.exit(1);
  }
}

// test route to make sure everything is working (accessed at GET http://localhost:8080/api)
router.get('/', function(req, res) {
    res.json({ message: 'hooray! welcome to our api!' });   
});

router.get('/ogdata', function(req, res) {
    res.json(ogdata);
});

// more routes for our API will happen here

// REGISTER OUR ROUTES -------------------------------
// all of our routes will be prefixed with /api
app.use('/api', router);

// START THE SERVER
// =============================================================================
app.listen(port);
fetch("http://rss.cnn.com/rss/cnn_topstories.rss");
fetch("http://feeds.foxnews.com/foxnews/most-popular");
console.log('Magic happens on port ' + port);


Gruntfile.js
var os = require('os');

module.exports = function(grunt) {
  grunt.initConfig({
    pkg: grunt.file.readJSON('package.json'),
    jshint: {
      all: ['<%= pkg.path.src.js %>/**/*.js']
    },
    concat: {
      cutup: {
        src: ['<%= pkg.path.cutup.js %>/plugins/**/*.js', '<%= pkg.path.src.js %>/**/*.js'],
        dest: '<%= pkg.path.cutup.js %>/scripts.js'
      }
    },
    uglify: {
      options: {
        mangle: {
          except: ['jQuery']
        },
        preserveComments: 'none',
        sourceMap: true
      },
      '<%= pkg.path.cutup.js %>/scripts.min.js': ['<%= pkg.path.cutup.js %>/scripts.js']
    },
    sass: {
      options: {
        sourceMap: true,
        outputStyle: 'compressed'
      },
      cutup: {
        files: {
          '<%= pkg.path.cutup.css %>/screen.css' : '<%= pkg.path.src.sass %>/screen.scss'
        }
      }
    },
    autoprefixer: {
      options: {
        // Task-specific options go here.
        browsers: ['last 2 versions', 'ie 8', 'ie 9'],
        map: true
      },
      cutup: {
        src: "<%= pkg.path.cutup.css %>/screen.css",
        dest: "<%= pkg.path.cutup.css %>/screen.css"
      }
    },
    combine_mq: {
      cutup: {
        expand: true,
        cwd: "<%= pkg.path.cutup.css %>",
        src: "**/*.css",
        dest: "<%= pkg.path.cutup.css %>",
        ext: ".css"
      }
    },
    cssmin: {
      cutup: {
        keepSpecialComments: true,
        expand: true,
        cwd: '<%= pkg.path.cutup.css %>/',
        src: ['*.css', '!*.min.css'],
        dest: '<%= pkg.path.cutup.css %>/',
        ext: '.min.css'
      }
    },  
    imagemin: {
      cutup: {
        options: {
          optimizationLevel: 7
        },
        files: [{
          expand: true,
          cwd: '<%= pkg.path.src.img %>/',
          src: '**/*.{jpg,png,gif,svg}',
          dest: '<%= pkg.path.cutup.img %>/'
        }]
      }
    },
    express: {
      dev: {
        options: {
          script: 'index.js',
          port:1337
        }
      },
      crawler:{
        options: {
          script: 'crawler.js',
          port:8080
        }        
      }
    },
    webfont: {
      icons: {
        src: '<%= pkg.path.src.font %>/icons/*.svg',
        dest: '<%= pkg.path.cutup.font %>',
        destCss: '<%= pkg.path.src.sass %>',
        options: {
          engine: os.platform() === 'win32' ? 'node' : 'fontforge',
          font: 'fontcustom',
          hashes: false,
          stylesheet: 'scss',
          relativeFontPath: '/library/fonts/',
          template: '<%= pkg.path.src.font %>/fontcustom.css',
          templateOptions: {
            baseClass: "",
            classPrefix: 'icon-',
            mixinPrefix: 'icon-'
          }
        }
      }
    },    
    clean: ["library/**/tmp"],
    watch: {      
      options: { livereload: true },
      scripts: {
        files: ['<%= pkg.path.cutup.js %>/plugins/*.js', '<%= pkg.path.src.js %>/*.js'],
        tasks: ['js']
      },
      css: {
        files: '<%= pkg.path.src.sass %>/**/*.scss',
        tasks: ['css']
      },
      img: {
        files: '<%= pkg.path.src.img %>/**/*.{jpg,gif,png,svg}',
        tasks: ['img']
      },
      html: {
        files: '**/*.html',
        tasks: []
      }
    },
    githooks: {
      all: {
        'post-merge': 'post-merge'
      }
    }
  });

  grunt.loadNpmTasks('grunt-contrib-watch');
  grunt.loadNpmTasks('grunt-contrib-jshint');
  grunt.loadNpmTasks('grunt-contrib-concat');
  grunt.loadNpmTasks('grunt-contrib-uglify');
  grunt.loadNpmTasks('grunt-sass');
  grunt.loadNpmTasks('grunt-combine-mq');
  grunt.loadNpmTasks('grunt-contrib-cssmin');
  grunt.loadNpmTasks('grunt-contrib-clean');
  grunt.loadNpmTasks('grunt-contrib-imagemin');
  grunt.loadNpmTasks('grunt-newer');
  grunt.loadNpmTasks('grunt-githooks');
  grunt.loadNpmTasks('grunt-npm-install');
  grunt.loadNpmTasks('grunt-autoprefixer');
  grunt.loadNpmTasks('grunt-webfont');
  grunt.loadNpmTasks('grunt-express-server');

  grunt.registerTask('js', ['jshint', 'concat', 'uglify', 'clean']);
  grunt.registerTask('css', ['sass', 'autoprefixer', 'combine_mq', 'cssmin', 'clean']);
  grunt.registerTask('img', ['newer:imagemin']);
  grunt.registerTask('post-merge', ['npm-install', 'default']);
  grunt.registerTask('server', ['express:dev', 'express:crawler', 'watch', ]);
  grunt.registerTask('default', ['js', 'css:development', 'img', 'server']);
}